### DISCUSSION TITLE
Three tricks with Dictionary

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#70] [General]
# Three tricks with Dictionary

https://steven-giesel.com/blogPost/53090b7c-d594-4b55-b513-60bcc7b36f35

**Comment by musmanrafiq** [comment]
Nice post, may be can you double check the method GetValueOrDefault<TKey> should be GetValueOrDefault<TKey, TValue>.

**Reply by linkdotnet** [reply]
Yes absolutely! Thanks - fixed

**Comment by jamescurran** [comment]
I really wish MSFT would add a ctor overload for Dictionary which would accept a Func<TKey, TValue> (a lambda which accept a TKey value and returns a TValue). Such that:

Dictionary<int, string> dictionary = new( _=> "not found")
{
    { 1, "foo" }
};

_ = dictionary[1];  // foo
_ = dictionary[2];  // not found

The current behavior could be implemented simply as:
Dictionary<int, string> dictionary = new( key => throw KeyNotFoundException ($"The given key '{key}' was not present in the dictionary.");

Or it could be used to simplify a cache:
Dictionary<int, string> cache = new( key => _repo.LookupValue(key));

**Reply by linkdotnet** [reply]
I like the last use case, especially the caching one.
At least for the code under your control you could do this:
```csharp
public class DefaultDictionary<TKey, TValue> : Dictionary<TKey, TValue>
{
    private readonly Func<TKey, TValue> defaultValueProvider;

    public DefaultDictionary(Func<TKey, TValue> defaultValueProvider)
    {
        this.defaultValueProvider = defaultValueProvider;
    }

    public new TValue this[TKey key]
    {
        get
        {
            return TryGetValue(key, out var value))
                ? value
                : defaultValueProvider(key);
        }
        set => base[key] = value;
    }
}
```

**Comment by LexaGV** [comment]
These advises clearly show how pathetic MS architectors are.  They think as a dictator: PROHIBIT EVERYTHING! Key is null? Kill him! Key doesn't exist? Cut programmer on pieces! Who the hell develops collections like that!?!? Total idiots.... And rather than these advises, I made YEARS ago normal Dictionary as it HAD to be done: with case insensitive keys, with default value when key doesn't exist, serialization, etc. I despise MS programmers for buggy, clumsy, narrow mind code - they must NOT touch base library as they are incompetent.

**Reply by linkdotnet** [reply]
Honest question: Aren't almost all languages do such behavior? If the key doesn't exist python also throws an error. Java does the same and go returns an error and the default value for the key.

**Reply by jamescurran** [reply]
If you got back to the old-school .NET collections, the Dictionary (which mapped an object to an object) would return a null if the key wasn't found.  But, since the generic DIctionary<> could have value-typed TValues, it can't return null.   (StringDictionary also returns null)

**Reply by Patrick-3000** [reply]
I am actually happy with the current behaviour. It makes you think and handle cases such as non-existing keys etc. And if you don't handle them, then at least you fail early instead of continuing with probably missing or incorrect data, which can lead to hard-to-find bugs.

**Comment by aathar** [comment]
I came across a great extensions library (that included the above points) and then lost it somewhere. Overall its an interesting read.

**Reply by linkdotnet** [reply]
Nice to hear! Microsoft is adding more and more such convenient functions to the framework itself.

**Comment by q34hfq34fq5374rqhdf0vih3435dd23234wkklo** [comment]
Hi, the intro sentence says "...I want to showcase two nice tricks..." think you meant three :)

**Reply by linkdotnet** [reply]
😀 Oh boy yeah! Thanks for that. I fixed that
